File: m4-h2/src/main/java/ru/yegorpilipenko/otus/spring/m4h2/service/GenreServiceImpl.java
package ru.yegorpilipenko.otus.spring.m4h2.service;

import com.netflix.hystrix.contrib.javanica.annotation.HystrixCommand;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import ru.yegorpilipenko.otus.spring.m4h2.model.Genre;
import ru.yegorpilipenko.otus.spring.m4h2.repository.BookRepository;

import java.util.Collection;

@Service
@RequiredArgsConstructor
public class GenreServiceImpl implements GenreService {

    private final BookRepository bookRepository;

    @Override
    @HystrixCommand(groupKey = "Genres", commandKey = "GetAllGenres")
    public Collection<Genre> getAll() {
        return bookRepository.findAllGenres();
    }

}

File: m2-h3/src/main/kotlin/ru/yegorpilipenko/otus/spring/m2h3/service/GenreServiceImpl.kt
package ru.yegorpilipenko.otus.spring.m2h3.service

import org.springframework.stereotype.Service
import ru.yegorpilipenko.otus.spring.m2h3.model.Book
import ru.yegorpilipenko.otus.spring.m2h3.model.Genre
import ru.yegorpilipenko.otus.spring.m2h3.repository.BookRepository

@Service
class GenreServiceImpl(val bookRepository: BookRepository): GenreService {

    override fun getAll(): Collection<Genre> = bookRepository.findAll().flatMap(Book::genres).toSet()

}
